File: backend-api-deprecated/src/types/judgment.ts
/**
 * 质量判定模块类型定义
 */

import { JudgmentResult } from '@prisma/client'

/**
 * 判定规则类型枚举
 */
export enum JudgmentRuleType {
  RANGE = 'RANGE', // 范围判定
  FORMULA = 'FORMULA', // 公式判定
  LOGIC = 'LOGIC' // 逻辑表达式判定
}

/**
 * 判定规则条件
 */
export interface JudgmentRuleCondition {
  type: JudgmentRuleType
  parameter?: string // 检测参数名称
  minValue?: number // 最小值（范围判定）
  maxValue?: number // 最大值（范围判定）
  formula?: string // 公式表达式（公式判定）
  logicExpression?: string // 逻辑表达式（逻辑判定）
  expectedResult?: JudgmentResult // 期望结果
}

/**
 * 判定规则配置
 */
export interface JudgmentRuleConfig {
  id: string
  name: string
  description?: string
  testItemType: string // 检测项类型
  conditions: JudgmentRuleCondition[] // 判定条件（AND 关系）
  priority: number // 优先级（数字越大优先级越高）
  isActive: boolean
  createdBy: string
  createdAt: Date
  updatedAt: Date
}

/**
 * 创建判定规则 DTO
 */
export interface CreateJudgmentRuleDto {
  name: string
  description?: string
  testItemType: string
  conditions: JudgmentRuleCondition[]
  priority?: number
}

/**
 * 更新判定规则 DTO
 */
export interface UpdateJudgmentRuleDto {
  name?: string
  description?: string
  conditions?: JudgmentRuleCondition[]
  priority?: number
  isActive?: boolean
}

/**
 * 判定规则查询参数
 */
export interface JudgmentRuleQuery {
  testItemType?: string
  isActive?: boolean
  page?: number
  pageSize?: number
}

/**
 * 执行质量判定 DTO
 */
export interface PerformJudgmentDto {
  sampleId: string
  ruleIds?: string[] // 指定使用的规则 ID，如果不指定则自动匹配
  performedBy: string
}

/**
 * 判定依据详情
 */
export interface JudgmentBasisDetail {
  ruleId: string
  ruleName: string
  conditionType: JudgmentRuleType
  parameter?: string
  actualValue?: number | string
  expectedRange?: { min?: number; max?: number }
  formula?: string
  calculatedValue?: number
  logicExpression?: string
  evaluationResult: boolean
  message: string
}

/**
 * 判定结果响应
 */
export interface JudgmentResponse {
  id: string
  sampleId: string
  result: JudgmentResult
  basis: string // JSON 字符串，包含判定依据详情
  basisDetails: JudgmentBasisDetail[] // 解析后的判定依据
  isAutomatic: boolean
  judgedBy: string
  judgedAt: Date
  reviewedBy?: string
  reviewedAt?: Date
}

/**
 * 人工复核判定 DTO
 */
export interface ReviewJudgmentDto {
  judgmentId: string
  newResult: JudgmentResult
  reason: string
  reviewedBy: string
}

/**
 * 判定历史记录
 */
export interface JudgmentHistory {
  id: string
  judgmentId: string
  sampleId: string
  previousResult: JudgmentResult
  newResult: JudgmentResult
  changeReason: string
  changedBy: string
  changedAt: Date
}

/**
 * 判定历史查询参数
 */
export interface JudgmentHistoryQuery {
  sampleId?: string
  judgmentId?: string
  page?: number
  pageSize?: number
}

/**
 * 批量判定 DTO
 */
export interface BatchJudgmentDto {
  sampleIds: string[]
  performedBy: string
}

/**
 * 批量判定结果
 */
export interface BatchJudgmentResult {
  total: number
  successful: number
  failed: number
  results: Array<{
    sampleId: string
    success: boolean
    judgment?: JudgmentResponse
    error?: string
  }>
}

File: backend-api/app/models/judgment.py
"""
质量判定模型

此模块定义了与 Prisma schema 兼容的质量判定相关 SQLAlchemy 模型。
包括：QualityJudgment、JudgmentRule、JudgmentHistory
"""
from sqlalchemy import Column, String, Integer, Boolean, DateTime, Enum as SQLEnum, ForeignKey, Text, JSON
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
import uuid
import enum
from app.models.base import Base


class JudgmentResult(str, enum.Enum):
    """判定结果枚举 - 匹配 Prisma schema"""
    QUALIFIED = "QUALIFIED"
    UNQUALIFIED = "UNQUALIFIED"
    PENDING = "PENDING"


class QualityJudgment(Base):
    """
    质量判定模型 - 兼容 Prisma schema
    
    表示样品的质量判定结果，包括判定结果、判定依据、判定人员等信息。
    """
    __tablename__ = 'quality_judgments'
    
    # 主键
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    
    # 样品关联（一对一关系）
    sampleId = Column(String, ForeignKey('samples.id', ondelete='CASCADE'), unique=True, nullable=False)
    
    # 判定信息
    result = Column(
        SQLEnum(JudgmentResult, name="JudgmentResult"),
        nullable=False,
        index=True
    )
    basis = Column(Text, nullable=False)  # 判定依据（JSON 格式）
    isAutomatic = Column(Boolean, default=True, nullable=False)
    version = Column(Integer, default=1, nullable=False)  # 乐观锁版本号
    
    # 审计字段
    judgedBy = Column(String, nullable=False)
    judgedAt = Column(DateTime, server_default=func.now(), nullable=False)
    reviewedBy = Column(String)
    reviewedAt = Column(DateTime)
    
    # 关系
    sample = relationship('Sample', back_populates='quality_judgment', uselist=False)
    history = relationship('JudgmentHistory', back_populates='judgment', cascade='all, delete-orphan')
    
    def __repr__(self):
        return f"<QualityJudgment(id={self.id}, sampleId={self.sampleId}, result={self.result})>"


class JudgmentRule(Base):
    """
    判定规则配置模型 - 兼容 Prisma schema
    
    表示质量判定的规则配置，包括规则名称、检测项类型、判定条件等。
    """
    __tablename__ = 'judgment_rules'
    
    # 主键
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    
    # 规则信息
    name = Column(String, nullable=False)
    description = Column(String)
    testItemType = Column(String, nullable=False, index=True)  # 检测项类型
    conditions = Column(JSON, nullable=False)  # 判定条件配置
    priority = Column(Integer, default=0, nullable=False)
    isActive = Column(Boolean, default=True, nullable=False, index=True)
    
    # 审计字段
    createdBy = Column(String, nullable=False)
    createdAt = Column(DateTime, server_default=func.now(), nullable=False)
    updatedAt = Column(
        DateTime,
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False
    )
    
    def __repr__(self):
        return f"<JudgmentRule(id={self.id}, name={self.name}, testItemType={self.testItemType})>"


class JudgmentHistory(Base):
    """
    判定历史记录模型 - 兼容 Prisma schema
    
    记录质量判定结果的变更历史，包括变更前后的结果、变更原因等。
    """
    __tablename__ = 'judgment_history'
    
    # 主键
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    
    # 判定关联
    judgmentId = Column(String, ForeignKey('quality_judgments.id', ondelete='CASCADE'), nullable=False, index=True)
    sampleId = Column(String, nullable=False, index=True)
    
    # 变更信息
    previousResult = Column(SQLEnum(JudgmentResult, name="JudgmentResult"), nullable=False)
    newResult = Column(SQLEnum(JudgmentResult, name="JudgmentResult"), nullable=False)
    changeReason = Column(Text, nullable=False)
    
    # 审计字段
    changedBy = Column(String, nullable=False)
    changedAt = Column(DateTime, server_default=func.now(), nullable=False, index=True)
    
    # 关系
    judgment = relationship('QualityJudgment', back_populates='history')
    
    def __repr__(self):
        return f"<JudgmentHistory(id={self.id}, judgmentId={self.judgmentId}, previousResult={self.previousResult}, newResult={self.newResult})>"
